### PR TITLE
Hard limit the number of results returned by a search

### LINKED ISSUE #2133: Hard limit when searching documents
To avoid issues like this one: https://github.com/meilisearch/meilisearch/issues/1708, we need to fix an hard limit during the search.

WIP on the product side.

### PR DESCRIPTION
**Pull request by Kerollmops**
This PR fixes #2133 by hard-limiting the number of results that a search request can return at any time. I would like the guidance of @MarinPostma to test that, should I use a mocking test here? Or should I do anything else?

I talked about touching the _nb_hits_ value with @qdequele and we concluded that it was not correct to do so.

Could you please confirm that it is the right place to change that?

### CODE REVIEW

**Comment by gmourier** [comment]
> I talked about touching the nb_hits value with @qdequele and we concluded that it was not correct to do so.

Curious to hear your thoughts on this! What were the reasons for making this choice?

**Comment by MarinPostma** [comment]
I don't think that this is how it should be done.

what if I ask for 1 document at offset 10000?

I think we need to pass the following params to milli:

```
offset = min(offset, HARD_RESULT_LIMIT);
limit = min(limit, HARD_RESULT_LIMIT - OFFSET);
```

wdyt?

**Comment by Kerollmops** [comment]
Ahha yeah, I'm dumb, you are absolutely right. That's probably due to jet lag. I will update my PR accordingly. Thanks!

File: meilisearch-lib/src/index/search.rs
Comment (MarinPostma) [inline meilisearch-lib/src/index/search.rs:107]:
```suggestion
        let offset = query.offset.unwrap_or(0).min(HARD_RESULT_LIMIT));
```

File: meilisearch-lib/src/index/search.rs
Comment (MarinPostma) [inline meilisearch-lib/src/index/search.rs:108]:
```suggestion
        let limit = query.limit.min(HARD_RESULT_LIMIT.saturating_sub(offset));
```

the saturating sub is not strictly necessary since we already clamped `offset` to `HARD_RESULT_LIMIT`, but I don't see any downsides, so cool

File: meilisearch-lib/src/index/search.rs
Comment (MarinPostma) [inline meilisearch-lib/src/index/search.rs:1]:
```suggestion
```

**Review by MarinPostma — CHANGES_REQUESTED** [review]

**Comment by MarinPostma** [comment]
@Kerollmops, for testability, you could in fact do a mock here, but maybe the simple option would be to make a function like `compute_offset_limit(offset: usize, limit: usize) -> (usize, uszie)`, and unit test that function instead, wdyt?

If we wanted to make integration tests, having a way to lower the value of `MAX_RESULT_LIMIT` would be useful so we don't have to create search results with 1000 hits, but who should own this setting is another question...

File: meilisearch-lib/src/index/search.rs
Comment (Kerollmops) [reply meilisearch-lib/src/index/search.rs:107]: I find it way easier to use the `cmp::min` function like I do than using the `usize::min` method.

File: meilisearch-lib/src/index/search.rs
Comment (Kerollmops) [reply meilisearch-lib/src/index/search.rs:108]: Yeah you are right about the `saturating_sub` uselessness and I'll keep it :)

**Review by curquiza — COMMENTED** [review]
Ok for me ✅

Wait for the final Marin's approval for the merge

**Review by MarinPostma — APPROVED** [review]

**Pull request merged**
